modules/private.php

```diff
@@ -0,0 +1,341 @@
+		$is_admin_post_request = ( 'admin-post.php' === $pagenow );
+
+		if ( is_super_admin() ) {
+			return;
+		}
+		
+		
+		// Make sure we are in the right code path, if not bail now.
+		
+		if ( ! is_admin() || ( ! $is_ajax_request && ! $is_admin_post_request ) ) {
+			return;
+		}
+
+		if ( ! Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) {
+			wp_die();
+		}
+	}
+
+	/**
+	 * Prevents ajax requests on private blogs for users who don't have permissions
+	 *
+	 * @param string    $action The Ajax nonce action.
+	 * @param false|int $result The result of the nonce check.
+	 */
+	static function private_blog_ajax_nonce_check( $action, $result ) {
+		global $wpdb;
+
+		$this_current_user = wp_get_current_user();
+
+		if ( is_super_admin() ) {
+			return;
+		}
+
+		if ( 1 !== $result && 2 !== $result ) {
+			return;
+		}
+
+		if ( 'find-posts' !== $action && 'internal-linking' !== $action ) {
+			return;
+		}
+
+		// Make sure we are in the right code path, if not bail now.
+		if ( ! is_admin() || ( ! defined( 'DOING_AJAX' ) || ! DOING_AJAX ) ) {
+			return;
+		}
+
+		if ( ! Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) {
+			wp_die( -1 );
+		}
+	}
+
+	/**
+	 * Disables WordPress Rest API for external requests
+	 */
+	static function disable_rest_api() {
+		global $wpdb;
+		$this_current_user = wp_get_current_user();
+
+		if ( is_user_logged_in() && ( is_super_admin() || Jetpack_Private::is_private_blog_user( $wpdb->blogid, $this_current_user ) ) ) {
+			return;
+		}
+
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return new WP_Error( 'private_site', __( 'This site is private.', 'jetpack' ), 403 );
+		}
+	}
+
+	static function module_override( $modules ) {
+		$disabled_modules = array(
+			'publicize',
+			'sharedaddy',
+			'subscriptions',
+			'json-api',
+			'enhanced-distribution',
+			'google-analytics',
+			'photon',
+			'sitemaps',
+			'verification-tools',
+			'wordads',
+		);
+
+		foreach ( $disabled_modules as $module_slug ) {
+			$found = array_search( $module_slug, $modules );
+			if ( false !== $found ) {
+				unset( $modules[ $found ] );
+			}
+		}
+
+		return $modules;
+	}
+
+	/**
+	 * Show an error when the blog_public option is updated
+	 */
+	static function private_update_option_blog_public() {
+		if ( function_exists( 'add_settings_error') ) {
+			add_settings_error( 'general', 'setting_not_updated', sprintf( __( 'This setting is ignored because you <a href="%s">made your site private</a>', 'jetpack' ), admin_url( 'admin.php?page=jetpack' ) . '#/traffic?term=private' ), 'error' );
+		}
+	}
+
+	/**
+	 * Adds a message to the 'At a Glance' dashboard widget.
+	 */
+	static function add_private_dashboard_glance_items( $content ) {
+		return $content . '<br><br>' . __( 'This site is currently Private', 'jetpack' );
+	}
+
+	/**
+	 * Changes a blog to private
+	 */
+	static function make_blog_private() {
+		update_option( 'blog_public', -1 );
```

Comment:
You're right - it's only on WordPress.com that `blog_public` takes _-1_ , not on core. Should we instead make it _0_ here, even though that's not the best representation of the blog's state?

Reply:
On WPCOM we make it -1. I wonder if we should leave the value of blog_public unchanged when the user activates the module, and find another way to expose the value publicly.

Reply:
I don't have an answer :) I want to sure we understand the impacts of the various choices, though, so we can figure out that best answer.

It could be using `blog_public=0` and finding another way to expose our state is best.
It could be using `blog_public=-1` and making sure core knows what to do with that (via more filters on those functions I mentioned above) is best.

What do other private site plugins do?

Can we talk to core and get a common method in place? (This is what we did for post geo data way back in the day, though that was much simpler situation: a very new feature without any backward compatibility concerns.)

Reply:
I _think_ core used to support -1 for private blogs and it was removed. There's certainly some code in wp-admin for exposing those options.

I'd rather not use `blog_public=0` as that has an explicit meaning in WordPress, which is different to what this module does.

We sometimes recommend people use @Viper007Bond's Registered User's Only plugin: https://plugins.trac.wordpress.org/browser/registered-users-only/. From what I can see this doesn't use `blog_public`.

IMO it would be simpler to avoid changing blog_public altogether.

Reply:
What is your thought about supporting the Private option in Calypso? For the Jetpack site in Calypso, `/settings/general/{siteSlug}` shows only the Public and Hidden options. Should we show Private here? If yes, would it help having `blog_public=-1`?

<img width="921" alt="Screenshot 2019-03-14 at 8 36 23 PM" src="https://user-images.githubusercontent.com/1269602/54367832-13583800-4699-11e9-8407-141e98ad4b40.png">

If this is not applicable, or not a priority, then I agree with @scruffian that not changing `blog_public ` could be the better option.

Reply:
I think ideally we should use the same module for private sites in WPCOM and in Jetpack. WPCOM uses the blog_private option, but as @mdawaffe points out this is problematic with core as sometimes core expects the value to be 0 or 1, not -1. Because of the I think we shouldn't rely on blog_public at all in either Jetpack or WPCOM...

Reply:
I've removed modifications to `blog_public` and introduced a change to the Sitemap module - please check (86d06ed)[https://github.com/Automattic/jetpack/pull/10333/commits/86d06ed10e5ee1febdfc7a2332a21d35f1facf9c]

Reply:
> I think ideally we should use the same module for private sites in WPCOM and in Jetpack. 

I think so too. It will take a lot of work to get there, as you know :)